### PR TITLE
Transactional json logging

### LINKED ISSUE #1799: Add transactional json logging for dbt invocations
### Describe the feature
[This PR](https://github.com/fishtown-analytics/dbt/pull/1791) adds a json logger to the dbt project. When json logging is used, dbt should emit "transactional" log lines that report on the state of the run. From these logs alone, external processes should be able to paint a UI with a superset of the stdout logging that dbt presents on the CLI. This includes:
 - The current state of the dbt project (`internal`, `parsing`, `running`)
 - The current state of the resources that have run, or are running:
  - state: `running`, `complete`
  - status: `running`, `cancelled`, `success`, `fail`, `error`, etc. 
  - status_message: (the thing we print on stdout at resource completion)
  - timing: (`started_at`, `finished_at`)
  - error: if an error occurred in the scope of a resource, the error (as a string)

Implementation notes:
 - log lines that occur during each of `internal`, `parsing`, and `running` should be tagged with an attribute, `run_state` (or similar), so they can be collected and rendered together
  - `internal` is for log lines that occur outside the scope of `parsing` and `running`
- these log lines should be easily identifiable from the log stream, eg with the `channel`  attribute, or similar
- we can either cram this information into the `extra` dict, or we can represent this data some other way if that makes more sense.

### PR DESCRIPTION
**Pull request by beckjake**
Fixes #1799 

This is based on #1801, so once that is merged I'll rebase this against dev/louisa-may-alcott.

### CODE REVIEW

**Comment by beckjake** [comment]
> how does the exc_info field get populated?

If you log a message with `exc_info=True` in an `except` block (and maybe in a `finally` block if there's an active exception), logging and logbook both insert that.

>can we better structure the field values around success/error states? It looks like model_state can either be passed, or a literal error message. If possible, it would be good to present a field which has a value in {success, error, skipped} that can be used in conjunction with a free-text field which describes the error (if one occurs).

Yeah, we can do this, I think. Part of the problem is that dbt's `error`, `status`, etc fields in results are a bit ad-hoc. I'm probably being generous with that. We should probably fix the underlying data structures instead of trying to fix it on the logger side, which is a bit involved.

**Comment by beckjake** [comment]
I'm going to resolve that error/status issue by just calling `str` on things that might not be strings, instead of cleaning up the underlying behavior - we can squash any issues with weird outputs as we find them.

File: core/dbt/node_runners.py [hunk 1]
```diff
@@ -70,6 +70,18 @@ def __init__(self, config, adapter, node, node_index, num_nodes):
...
+        if result.error:
+            return {'model_status': 'error', 'model_error': str(result.error)}
+        elif result.skip:
+            return {'model_status': 'skipped'}
+        elif result.fail:
+            return {'model_status': 'failed'}
+        elif result.warn:
+            return {'model_status': 'warn'}
+        else:
+            return {'model_status': 'passed'}
```
Comment (drewbanin) [inline core/dbt/node_runners.py]: can we change all of these from `model_status` to `node_status` for consistency? Many of these results will be generated from nodes which are not models.

File: core/dbt/node_runners.py [hunk 2]
```diff
@@ -436,6 +448,12 @@ def on_skip(self):
             'Freshness: nodes cannot be skipped!'
         )
 
+    def get_result_status(self, result) -> Dict[str, str]:
+        if result.error:
+            return {'model_status': 'error', 'model_error': str(result.error)}
```
Comment (drewbanin) [inline core/dbt/node_runners.py]: Let's also make this `node_error` instead of `model_error`

File: core/dbt/logger.py [hunk 3]
```diff
@@ -186,6 +199,107 @@ def process(self, record):
...
+        self.node_index = node_index
+        super().__init__()
+
+    def process(self, record):
+        keys = [
+            ('alias', 'node_alias'),
+            ('schema', 'node_schema'),
+            ('database', 'node_database'),
+            ('name', 'node_name'),
+            ('original_file_path', 'node_path')
```
Comment (drewbanin) [inline core/dbt/logger.py]: can we also include the resource type here?

**Pull request merged**